File: contribs/src/main/scala/io/delta/storage/BaseExternalLogStore.scala
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.storage

import java.io.FileNotFoundException
import java.net.URI
import java.nio.charset.StandardCharsets.UTF_8
import java.nio.file

import scala.util.control.NonFatal

import com.google.common.io.CountingOutputStream
import org.apache.commons.io.IOUtils
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.fs._

import org.apache.spark.SparkConf
import org.apache.spark.sql.delta.metering.DeltaLogging
import org.apache.spark.sql.delta.storage.HadoopFileSystemLogStore
import org.apache.spark.sql.delta.util.FileNames

abstract class BaseExternalLogStore(sparkConf: SparkConf, hadoopConf: Configuration)
  extends HadoopFileSystemLogStore(sparkConf, hadoopConf) with DeltaLogging {

  private val MILLIS_IN_DAY = 86400000

  ///////////////////////////////////////////////////////////////////////////
  // Public API Overrides
  ///////////////////////////////////////////////////////////////////////////

  override def isPartialWriteVisible(path: Path): Boolean = false

  /**
   * List files starting from `resolvedPath` (inclusive) in the same directory.
   */
  override def listFrom(path: Path): Iterator[FileStatus] = {
    logDebug(s"listFrom path: ${path}")
    val (fs, resolvedPath) = resolved(path)
    listFrom(fs, resolvedPath)
  }

  override def write(
      path: Path,
      actions: Iterator[String],
      overwrite: Boolean = false): Unit = {
    val (fs, resolvedPath) = resolved(path)

    // 1. If N.json already exists in the _delta_log, exit early
    if (fs.exists(resolvedPath)) {
      if (overwrite) {
        return
      } else {
        throw new java.nio.file.FileAlreadyExistsException(resolvedPath.toString)
      }
    }

    // 2. Ensure N-1.json exists in the _delta_log
    val prevVersionPathOpt = getPreviousVersionPath(resolvedPath)
    if (prevVersionPathOpt.isDefined && !fs.exists(prevVersionPathOpt.get)) {
      val prevVersionPath = prevVersionPathOpt.get
      val prevVersionEntryOpt = lookupInCache(fs, prevVersionPath)

      if (prevVersionEntryOpt.isEmpty) {
        logWarning(s"While trying to write $path, the preceding _delta_log entry " +
          s"$prevVersionPath was not found and neither was the preceding external entry. " +
          s"This means that file $prevVersionPath has been lost.")
      } else {
        if (prevVersionEntryOpt.get.isComplete) {
          logWarning(s"External entry for file $prevVersionPath is marked as complete, " +
            s"but file does not exist in the _delta_log.")
        }

        logDebug(s"Previous file $prevVersionPath doesn't exist in the _delta_log. Fixing now.")
        tryFixTransactionLog(fs, prevVersionEntryOpt.get, failureAcceptable = false)
      }
    }

    logDebug(s"Writing file: $path, $overwrite")

    // 3. Create temp file T(N)
    val tempPath = getTemporaryPath(resolvedPath)
    val fileLength = writeActions(fs, tempPath, actions)

    // 4. Commit to external cache entry E(N, complete = false)
    val logEntryMetadata = LogEntryMetadata(resolvedPath, fileLength, tempPath, isComplete = false)
    try {
      writeCache(fs, logEntryMetadata, overwrite)
    } catch {
      case e: Throwable =>
        logError(s"${e.getClass.getName}: $e")
        throw e
    }

    // At this point, the commit is recoverable and any future errors are not passed to the user

    try {
      // 5. Copy T(N) into target N.json
      copyFile(fs, tempPath, resolvedPath)

      // 6. Commit to external cache entry E(N, complete = true)
      writeCache(fs, logEntryMetadata.complete(), overwrite = true)
    } catch {
      case e: Throwable =>
        logWarning(s"${e.getClass.getName}: ignoring recoverable error: $e")
    }

    // 7. Delete all external cache E(*, complete = false) entries older than 1 day
    val tempFilesToDelete = deleteFromCacheAllOlderThan(
      fs, resolvedPath.getParent, System.currentTimeMillis() - MILLIS_IN_DAY)

    // 8. Delete the temp file T(*) for each deleted external cache entry older than 1 day
    tempFilesToDelete.foreach(fs.delete(_, false))
  }

  ///////////////////////////////////////////////////////////////////////////
  // Protected Methods
  ///////////////////////////////////////////////////////////////////////////

  /**
   * Write cache in exclusive way.
   *
   * @throws java.nio.file.FileAlreadyExistsException if path exists in cache and `overwrite` is
   *                                                  false
   */
  protected def writeCache(
      fs: FileSystem,
      logEntry: LogEntryMetadata,
      overwrite: Boolean = false): Unit

  /**
   * List the paths in the external cache that are lexicographically greater or equal to
   * (UTF-8 sorting) the given `resolvedPath`. The result should also be sorted by the file name.
   */
  protected def listFromCache(fs: FileSystem, resolvedPath: Path): Iterator[LogEntryMetadata]

  /**
   * Returns the LogEntryMetadata from the external cache with the matching resolvedPath if it
   * exists, else None
   */
  protected def lookupInCache(fs: FileSystem, resolvedPath: Path): Option[LogEntryMetadata]

  /**
   * Deletes all incomplete external cache LogEntryMetadata entries with modification times greater
   * than or equal to the given `expirationTime`, and returns their temp paths.
   */
  protected def deleteFromCacheAllOlderThan(
      fs: FileSystem,
      parentPath: Path,
      expirationTime: Long): Iterator[Path]

  ///////////////////////////////////////////////////////////////////////////
  // Private Helper Methods
  ///////////////////////////////////////////////////////////////////////////

  /**
   * Returns the path for the previous version N-1 given the input version N, if such a path exists
   */
  def getPreviousVersionPath(path: Path): Option[Path] = {
    if (!FileNames.isDeltaFile(path)) return None

    val currVersion = FileNames.deltaVersion(path)

    if (currVersion > 0) {
      val prevVersion = currVersion - 1
      val parentPath = path.getParent
      Some(FileNames.deltaFile(parentPath, prevVersion))
    } else {
      None
    }
  }

  /**
   * Write file with actions under a specific path.
   */
  private def writeActions(fs: FileSystem,
      path: Path,
      actions: Iterator[String]): Long = {
    logDebug(s"writeActions to: $path")
    val stream = new CountingOutputStream(fs.create(path, true))
    actions.map(_ + "\n").map(_.getBytes(UTF_8)).foreach(stream.write)
    stream.close()
    stream.getCount
  }

  /**
   * Copies file within filesystem.
   *
   * Ensures that the `src` file is either entirely copied to the `dst` file, or not at all.
   *
   * @param fs  reference to [[FileSystem]]
   * @param src path to source file
   * @param dst path to destination file
   */
  private def copyFile(fs: FileSystem, src: Path, dst: Path) {
    logDebug(s"copy file: $src -> $dst")
    val input_stream = fs.open(src)
    val output_stream = fs.create(dst, true)
    try {
      IOUtils.copy(input_stream, output_stream)
      output_stream.close()
    } finally {
      input_stream.close()
    }
  }

  /**
   * Merge two iterators of [[FileStatus]] into a single iterator ordered by file path name.
   * In case both iterators have [[FileStatus]]s for the same file path, keep the one from
   * `iterWithPrecedence` and discard that from `iter`.
   */
  private def mergeFileIterators(
      iter: Iterator[FileStatus],
      iterWithPrecedence: Iterator[FileStatus]
  ): Iterator[FileStatus] = {
    val result = (
        iter.map(f => (f.getPath, f)).toMap ++
        iterWithPrecedence.map(f => (f.getPath, f))
      )
      .values
      .toSeq
      .sortBy(_.getPath.getName)
    result.iterator
  }

  private def resolved(path: Path): (FileSystem, Path) = {
    val fs = path.getFileSystem(getHadoopConfiguration)
    val resolvedPath = stripUserInfo(fs.makeQualified(path))
    (fs, resolvedPath)
  }

  private def stripUserInfo(path: Path): Path = {
    val uri = path.toUri
    val newUri = new URI(
      uri.getScheme,
      null,
      uri.getHost,
      uri.getPort,
      uri.getPath,
      uri.getQuery,
      uri.getFragment
    )
    new Path(newUri)
  }

  /**
   * Best-effort at ensuring consistency on file system according to the external cache.
   *
   * Tries to copy a LogEntryMetadata's temp file into the target _delta_log file, if it does not
   * exist already, and then commit to the external cache.
   *
   * Will retry at most 3 times.
   *
   * @return the correct FileStatus from which to read the entry's data
   */
  private def tryFixTransactionLog(
      fs: FileSystem,
      entry: LogEntryMetadata,
      failureAcceptable: Boolean): FileStatus = {
    logDebug(s"Try to fix: ${entry.path}")
    val completedEntry = entry.complete()
    var fileCopied = false

    for (i <- 0 until 3) {
      try {
        if (!fileCopied && !fs.exists(entry.path)) {
          copyFile(fs, entry.tempPath, entry.path)
          fileCopied = true
        }

        // Since overwrite is true, we do not expect any FileAlreadyExistsExceptions
        writeCache(fs, completedEntry, overwrite = true)

        return completedEntry.asFileStatus(fs)
      } catch {
        case NonFatal(e) =>
          logWarning(s"${e.getClass.getName}: Ignoring error while fixing. Iteration $i of 3. $e")
      }
    }

    if (failureAcceptable) {
      // We weren't able to commit to the external cache, but we can still read from the temp file
      entry.tempPathAsFileStatus(fs)
    } else {
      throw new RuntimeException(s"Unable to fix the transaction log for path ${entry.path}")
    }
  }

  /**
   * Returns path stripped user info.
   */
  protected def getPathKey(resolvedPath: Path): Path = {
    stripUserInfo(resolvedPath)
  }

  /**
   * Generate temporary path for TransactionLog.
   */
  protected def getTemporaryPath(path: Path): Path = {
    val uuid = java.util.UUID.randomUUID().toString
    new Path(s"${path.getParent}/.temp/${path.getName}.$uuid")
  }

  /**
   * List files starting from `resolvedPath` (inclusive) in the same directory, which merges
   * the file system list and the db list
   */
  protected def listFrom(
      fs: FileSystem,
      resolvedPath: Path,
      useCache: Boolean = true): Iterator[FileStatus] = {
    val parentPath = resolvedPath.getParent
    if (!fs.exists(parentPath)) {
      throw new FileNotFoundException(s"No such file or directory: $parentPath")
    }

    val listedFromFs =
      fs.listStatus(parentPath)
        .filter(path => !path.getPath.getName.endsWith(".temp"))
        .filter(path => path.getPath.getName >= resolvedPath.getName)

    if (!useCache) {
      return listedFromFs.iterator
    }

    val listedFromDB = listFromCache(fs, resolvedPath)
      .toList
      .map { entry =>
        if (!entry.isComplete) {
          tryFixTransactionLog(fs, entry, failureAcceptable = true)
        } else {
          entry.asFileStatus(fs)
        }
      }

    // for debug
    listedFromFs.iterator
      .map(entry => s"fs item: ${entry.getPath}")
      .foreach(x => logDebug(x))

    listedFromDB.iterator
      .map(entry => s"db item: ${entry.getPath}")
      .foreach(x => logDebug(x))
    // end debug

    mergeFileIterators(listedFromDB.iterator, listedFromFs.iterator)
  }
}

class CachedFileStatus(
    length: Long,
    isdir: Boolean,
    block_replication: Int,
    block_size: Long,
    modification_time: Long,
    path: Path
) extends FileStatus(length, isdir, block_replication, block_size, modification_time, path)

/**
 * The file metadata to be stored in the external db
 */
case class LogEntryMetadata(
    path: Path,
    length: Long,
    tempPath: Path,
    isComplete: Boolean,
    modificationTime: Long = System.currentTimeMillis()) {

  def complete(): LogEntryMetadata = {
    LogEntryMetadata(this.path, this.length, this.tempPath, isComplete = true)
  }

  def tempPathAsFileStatus(fs: FileSystem): FileStatus = {
    new CachedFileStatus(
      length,
      false,
      1,
      fs.getDefaultBlockSize(tempPath),
      modificationTime,
      tempPath
    )
  }

  def asFileStatus(fs: FileSystem): FileStatus = {
    new CachedFileStatus(
      length,
      false,
      1,
      fs.getDefaultBlockSize(path),
      modificationTime,
      path
    )
  }
}

File: storage-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
/*
 * Copyright (2021) The Delta Lake Project Authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.delta.storage;

import java.net.URI;
import java.net.URISyntaxException;
import java.nio.charset.StandardCharsets;
import io.delta.storage.HadoopFileSystemLogStore;
import org.apache.commons.io.IOUtils;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FSDataOutputStream;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.commons.collections.iterators.FilterIterator;

import java.io.IOException;
import java.util.Iterator;
import java.util.Optional;
import java.util.regex.Pattern;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
* Used so that we can use an external store child implementation
* to provide the mutual exclusion that the cloud store,
* e.g. s3, is missing.
*/
public abstract class BaseExternalLogStore extends HadoopFileSystemLogStore {
    private static final Logger LOG = LoggerFactory.getLogger(BaseExternalLogStore.class);

    ////////////////////////
    // Public API Methods //
    ////////////////////////

    public BaseExternalLogStore(Configuration hadoopConf) {
        super(hadoopConf);
        deltaFilePattern = Pattern.compile("\\d+\\.json");
    }

    @Override
    public Iterator<FileStatus> listFrom(Path path, Configuration hadoopConf) throws IOException {
        final FileSystem fs = path.getFileSystem(hadoopConf);
        final Path resolvedPath = stripUserInfo(fs.makeQualified(path));
        final Path tablePath = getTablePath(resolvedPath);
        final Optional<ExternalCommitEntry> entry = getLatestExternalEntry(tablePath);
        if (entry.isPresent()) {
            fixDeltaLog(fs, entry.get());
        }

        // This is predicated on the storage system providing consistent listing
        // If there was a recovery performed in the `fixDeltaLog` call, then some temp file
        // was just copied into some N.json in the delta log. Because of consistent listing,
        // the `super.listFrom` is guaranteed to see N.json.
        return super.listFrom(path, hadoopConf);
    }

    @Override
    public void write(
            Path path,
            Iterator<String> actions,
            Boolean overwrite,
            Configuration hadoopConf) throws IOException {
        final FileSystem fs = path.getFileSystem(hadoopConf);
        final Path resolvedPath = stripUserInfo(fs.makeQualified(path));

        if (overwrite) {
            writeActions(fs, path, actions);
            return;
        }

        final Path tablePath = getTablePath(resolvedPath);
        if (isDeltaFile(path)) {
            final long version = deltaVersion(path);
            if (version > 0) {
                final long prevVersion = version - 1;
                final Path prevPath = deltaFile(tablePath, prevVersion);
                final Optional<ExternalCommitEntry> entry = getExternalEntry(tablePath, prevPath);
                if (entry.isPresent()) {
                    fixDeltaLog(fs, entry.get());
                } else {
                    if (!fs.exists(prevPath)) {
                        throw new java.nio.file.FileSystemException(
                            String.format("previous commit %s doesn't exist", prevPath)
                        );
                    }
                }
            } else {
                final Optional<ExternalCommitEntry> entry = getExternalEntry(tablePath, path);
                if (entry.isPresent()) {
                    if (entry.get().complete && !fs.exists(path)) {
                        throw new java.nio.file.FileSystemException(
                            String.format(
                                "Old entries for %s still exist in the database", tablePath
                            )
                        );
                    }
                }
            }
        }

        final String tempPath = createTemporaryPath(resolvedPath);
        final ExternalCommitEntry entry = new ExternalCommitEntry(
            tablePath,
            resolvedPath.getName(),
            tempPath,
            false, // not complete
            null // commitTime
        );

        writeActions(fs, entry.absoluteTempPath(), actions);
        putExternalEntry(entry, false); // overwrite=false

        try {
            writeCopyTempFile(fs, entry.absoluteTempPath(), resolvedPath);
            writePutCompleteDbEntry(entry);
        } catch (Throwable e) {
            LOG.info(
                "{}: ignoring recoverable error: {}", e.getClass().getSimpleName(), e
            );
        }
    }

    @Override
    public Boolean isPartialWriteVisible(Path path, Configuration hadoopConf) {
        return false;
    }

    /////////////////////////////////////////////////////////////
    // Protected Members (for interaction with external store) //
    /////////////////////////////////////////////////////////////

    /**
     * Write file with actions under a specific path.
     */
    protected void writeActions(
        FileSystem fs,
        Path path,
        Iterator<String> actions
    ) throws IOException {
        LOG.debug("writeActions to: {}", path);
        FSDataOutputStream stream = fs.create(path, true);
        while(actions.hasNext()) {
            byte[] line = String.format("%s\n", actions.next()).getBytes(StandardCharsets.UTF_8);
            stream.write(line);
        }
        stream.close();
    }

    /**
     * Generate temporary path for TransactionLog.
     */
    protected String createTemporaryPath(Path path) {
        String uuid = java.util.UUID.randomUUID().toString();
        return String.format(".tmp/%s.%s", path.getName(), uuid);
    }

    /**
     * Returns the base table path for a given Delta log entry located in
     * e.g. input path of $tablePath/_delta_log/00000N.json would return $tablePath
     */
    protected Path getTablePath(Path path) {
        return path.getParent().getParent();
    }

    /**
     * Write to external store in exclusive way.
     *
     * @throws java.nio.file.FileAlreadyExistsException if path exists in cache and `overwrite` is
     *                                                  false
     */
    abstract protected void putExternalEntry(
        ExternalCommitEntry entry,
        boolean overwrite) throws IOException;

    /**
     * Return external store entry corresponding to delta log file with given `tablePath` and
     * `fileName`, or `Optional.empty()` if it doesn't exist.
     *
     * @param tablePath TODO
     * @param jsonPath TODO
     */
    abstract protected Optional<ExternalCommitEntry> getExternalEntry(
        Path tablePath,
        Path jsonPath) throws IOException;

    /**
     * Return the latest external store entry corresponding to the delta log for given `tablePath`,
     * or `Optional.empty()` if it doesn't exist.
     *
     * @param tablePath TODO
     */
    abstract protected Optional<ExternalCommitEntry> getLatestExternalEntry(
        Path tablePath) throws IOException;

    //////////////////////////////////////////////////////////
    // Protected Members (for error injection during tests) //
    //////////////////////////////////////////////////////////

    /**
     * Wrapper for `copyFile`, called by the `write` method.
     */
    protected void writeCopyTempFile(FileSystem fs, Path src, Path dst) throws IOException {
        copyFile(fs, src, dst);
    }

    /**
     * Wrapper for `putExternalEntry`, called by the `write` method.
     */
    protected void writePutCompleteDbEntry(ExternalCommitEntry entry) throws IOException {
        putExternalEntry(entry.asComplete(), true);
    }

    /**
     * Wrapper for `copyFile`, called by the `fixDeltaLog` method.
     */
    protected void fixDeltaLogCopyTempFile(FileSystem fs, Path src, Path dst) throws IOException {
        copyFile(fs, src, dst);
    }

    /**
     * Wrapper for `putExternalEntry`, called by the `fixDeltaLog` method.
     */
    protected void fixDeltaLogPutCompleteDbEntry(ExternalCommitEntry entry) throws IOException {
        putExternalEntry(entry.asComplete(), true);
    }

    ////////////////////
    // Helper Methods //
    ////////////////////

    /**
     * Method for assuring consistency on filesystem according to the external cache.
     * Method tries to rewrite TransactionLog entry from temporary path if it does not exist.
     */
    private void fixDeltaLog(FileSystem fs, ExternalCommitEntry entry) throws IOException {
        if (entry.complete) {
            return;
        }
        int retry = 0;
        boolean copied = false;
        while (true) {
            LOG.debug("trying to fix: {}", entry.fileName);
            try {
                if (!copied && !fs.exists(entry.absoluteJsonPath())) {
                    fixDeltaLogCopyTempFile(fs, entry.absoluteTempPath(), entry.absoluteJsonPath());
                    copied = true;
                }
                fixDeltaLogPutCompleteDbEntry(entry);
                LOG.info("fixed {}", entry.fileName);
                return;
            } catch(Throwable e) {
                LOG.info("{}: {}", e.getClass().getSimpleName(), e);
                if (retry >= 3) {
                    throw e;
                }
            }
            retry += 1;
        }
    }

   /**
    * Copies file within filesystem
    * @param fs reference to [[FileSystem]]
    * @param src path to source file
    * @param dst path to destination file
    */
    private void copyFile(FileSystem fs, Path src, Path dst) throws IOException {
        LOG.debug("copy file: {} -> {}", src, dst);
        FSDataInputStream input_stream = fs.open(src);
        FSDataOutputStream output_stream = fs.create(dst, true);
        try {
            IOUtils.copy(input_stream, output_stream);
            output_stream.close();
        } finally {
            input_stream.close();
        }
    }

    /**
     * Returns path stripped user info.
     */
    private Path stripUserInfo(Path path) {
        final URI uri = path.toUri();

        try {
            final URI newUri = new URI(
                uri.getScheme(),
                null, // userInfo
                uri.getHost(),
                uri.getPort(),
                uri.getPath(),
                uri.getQuery(),
                uri.getFragment()
            );

            return new Path(newUri);
        } catch (URISyntaxException e) {
            // Propagating this URISyntaxException to callers would mean we would have to either
            // include it in the public LogStore.java interface or wrap it in an
            // IllegalArgumentException somewhere else. Instead, catch and wrap it here.
            throw new IllegalArgumentException(e);
        }
    }

    // TODO - use java version of FileNames utils?
    private Pattern deltaFilePattern;

    private boolean isDeltaFile(Path path) {
        return deltaFilePattern.matcher(path.getName()).matches();
    }

    private Path deltaFile(Path tablePath, long version) {
        return new Path(tablePath, String.format("_delta_log/%020d.json", version));
    }

    private long deltaVersion(Path path) {
        return Long.parseLong(path.getName().split("\\.")[0]);
    }

}
